### PR TITLE
Add transactions to file writes and restore files to previous state when transactions fail

### PR DESCRIPTION
**Pull request by xxfast**
Addresses #91

- [x] Add missing documentation and labelled params
- [x] Update file codec to write to a temp file and move once completed
- [x] Add test cases

### CODE REVIEW

**Comment by SultanArshad** [comment]
@xxfast Thanks for the fix any update on this when it was going to be in production thanks.

**Comment by xxfast** [comment]
Hi @SultanArshad planning this for 0.9.0 release. Stay tuned

**Comment by xxfast** [comment]
Hi @SultanArshad This is ready for review :)

File: kstore-file/src/commonTest/kotlin/io/github/xxfast/kstore/file/FileCodecTests.kt [hunk 1]
```diff
@@ -37,23 +38,35 @@ class FileCodecTests {
...
+  @Test
+  fun testTransactionalEncode() = runTest {
+    codec.encode(listOf(MYLO))
+
+    // Encoder will fail half way through
+    assertFailsWith<NotImplementedError> { codec.encode(listOf(MYLO, KAT)) }
+
+    // The original file should not be touched
+    val actual: List<Pet>? = codec.decode()
+    assertEquals(listOf(MYLO), actual)
```
Comment (SultanArshad) [inline kstore-file/src/commonTest/kotlin/io/github/xxfast/kstore/file/FileCodecTests.kt:64]: Nice work @xxfast! I’m wondering what happens if a race condition occurs and two different agents/threads start writing to a file at the same time, but one of them gets canceled or stopped. Does this revert the file back to an older version that was not modified? What happens to the version of the file being updated by the other agent/thread?

I suggest there should be a way to check if someone is modifying the file, and the other process should have to wait or be queued until the current modification is finished.

Comment (xxfast) [reply kstore-file/src/commonTest/kotlin/io/github/xxfast/kstore/file/FileCodecTests.kt:64]: There should really be a single store for a single file, but if there's an external process that is writing to the same file - and if the store write was cancelled, it should delete the temp file and leave the original file untouched

**Pull request merged**